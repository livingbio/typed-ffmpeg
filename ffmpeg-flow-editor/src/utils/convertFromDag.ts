import { Node, Edge } from 'reactflow';
import {
  FilterNode,
  InputNode,
  OutputNode,
  GlobalNode,
  VideoStream,
  AudioStream,
  OutputStream,
} from '../types/dag';
import { predefinedFilters } from '../types/ffmpeg';
import { EdgeType, EdgeData } from '../types/edge';

export function convertFromDag(dag: GlobalNode): { nodes: Node[]; edges: Edge[] } {
  const nodes: Node[] = [];
  const edges: Edge[] = [];
  const nodeMap = new Map<string, string>(); // Maps DAG node to React Flow node ID

  // Helper function to create a unique ID for a node
  const createNodeId = (node: FilterNode | InputNode | OutputNode) => {
    if (nodeMap.has(node.toString())) {
      return nodeMap.get(node.toString())!;
    }
    if (node instanceof InputNode) return 'input-0';
    if (node instanceof OutputNode) {
      const index = nodes.filter(n => n.data.filterType === 'output').length;
      return `output-${index}`;
    }
    // For filter nodes, use a consistent ID format
    return `${node.name}-${nodes.filter(n => n.data.filterName === node.name).length}`;
  };

  // Helper function to get edge type from stream
  const getEdgeType = (stream: VideoStream | AudioStream): EdgeType => {
    if (stream instanceof VideoStream) return 'video';
    if (stream instanceof AudioStream) return 'audio';
    return 'av';
  };

  // First pass: Create all nodes
  const processNode = (node: FilterNode | InputNode | OutputNode) => {
    if (nodeMap.has(node.toString())) return;

    const nodeId = createNodeId(node);
    nodeMap.set(node.toString(), nodeId);

    if (node instanceof InputNode) {
      nodes.push({
        id: nodeId,
        type: 'input',
        position: { x: 100, y: 100 }, // React-specific property
        data: {
          label: 'Input',
          filterType: 'input',
          filterString: '[0:v]',
          parameters: {},
          handles: {
            inputs: [],
            outputs: [{ id: 'output-0', type: 'av' }],
          },
        },
      });
    } else if (node instanceof OutputNode) {
      nodes.push({
        id: nodeId,
        type: 'output',
        position: { x: 800, y: 100 }, // React-specific property
        data: {
          label: 'Output',
          filterType: 'output',
          filterString: '[outv]',
          parameters: {},
          handles: {
            inputs: node.inputs.map((_, i) => ({ id: `input-${i}`, type: 'av' })),
            outputs: [],
          },
          filename: node.filename,
        },
      });
    } else if (node instanceof FilterNode) {
      const filter = predefinedFilters.find((f) => f.name === node.name);
      if (!filter) return;

      nodes.push({
        id: nodeId,
        type: 'filter',
        position: { x: Math.random() * 500 + 200, y: Math.random() * 300 + 100 }, // React-specific property
        data: {
          label: node.name,
          filterType: 'filter',
          filterName: node.name,
          parameters: node.kwargs,
          handles: {
            inputs: node.input_typings.map((type, index) => ({
              id: `input-${index}`,
              type: type.value === 'audio' ? 'audio' : type.value === 'video' ? 'video' : 'av',
            })),
            outputs: node.output_typings.map((type, index) => ({
              id: `output-${index}`,
              type: type.value === 'audio' ? 'audio' : type.value === 'video' ? 'video' : 'av',
            })),
          },
        },
      });
    }

    // Process all input streams recursively
    node.inputs.forEach((input) => {
      if (input instanceof VideoStream || input instanceof AudioStream) {
        processNode(input.node);
      }
    });
  };

  // Process all output nodes from the GlobalNode
  dag.inputs.forEach((outputStream) => {
    if (outputStream instanceof OutputStream) {
      processNode(outputStream.node);
    }
  });

  // Second pass: Create edges
  const createEdges = (node: FilterNode | InputNode | OutputNode) => {
    const sourceId = nodeMap.get(node.toString());
    if (!sourceId) return;

    node.inputs.forEach((input, index) => {
      if (input instanceof VideoStream || input instanceof AudioStream) {
        const targetId = nodeMap.get(input.node.toString());
        if (!targetId) return;

        const edgeType = getEdgeType(input);
        const edge: Edge<EdgeData> = {
          id: `edge-${targetId}-output-${input.index || 0}-${sourceId}-input-${index}`,
          source: targetId,
          target: sourceId,
          sourceHandle: `output-${input.index || 0}`,
          targetHandle: `input-${index}`,
          type: 'smoothstep',
          style: { stroke: edgeType === 'video' ? '#f44336' : edgeType === 'audio' ? '#2196f3' : '#9c27b0' },
          data: { type: edgeType },
        };
        edges.push(edge);
      }
    });
  };

  // Create edges for all nodes
  dag.inputs.forEach((outputStream) => {
    if (outputStream instanceof OutputStream) {
      createEdges(outputStream.node);
    }
  });

  return { nodes, edges };
} 