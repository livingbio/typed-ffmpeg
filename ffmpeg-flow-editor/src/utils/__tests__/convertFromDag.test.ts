import { describe, it, expect } from 'vitest';
import { convertFromDag } from '../convertFromDag';
import {
  GlobalNode,
  InputNode,
  OutputNode,
  FilterNode,
  VideoStream,
  AudioStream,
  OutputStream,
  StreamType,
} from '../../types/dag';

describe('convertFromDag', () => {
  it('should convert a simple input-output DAG to React Flow nodes and edges', () => {
    // Create a simple DAG: Input -> Output
    const inputNode = new InputNode('input.mp4');
    const outputNode = new OutputNode('output.mp4', [new VideoStream(inputNode)]);
    const dag = new GlobalNode([new OutputStream(outputNode)]);

    const result = convertFromDag(dag);
    expect(result).toMatchSnapshot();
  });

  it('should convert a DAG with filter nodes to React Flow nodes and edges', () => {
    // Create a DAG: Input -> Filter -> Output
    const inputNode = new InputNode('input.mp4');
    const filterNode = new FilterNode(
      'reverse',
      [new VideoStream(inputNode)],
      [new StreamType('video')],
      [new StreamType('video')]
    );
    const outputNode = new OutputNode('output.mp4', [new VideoStream(filterNode)]);
    const dag = new GlobalNode([new OutputStream(outputNode)]);

    const result = convertFromDag(dag);
    expect(result).toMatchSnapshot();
  });

  it('should handle multiple output nodes', () => {
    // Create a DAG with multiple outputs: Input -> Filter -> Output1, Output2
    const inputNode = new InputNode('input.mp4');
    const filterNode = new FilterNode(
      'split',
      [new VideoStream(inputNode)],
      [new StreamType('video')],
      [new StreamType('video'), new StreamType('video')]
    );
    const outputNode1 = new OutputNode('output1.mp4', [new VideoStream(filterNode, 0)]);
    const outputNode2 = new OutputNode('output2.mp4', [new VideoStream(filterNode, 1)]);
    const dag = new GlobalNode([
      new OutputStream(outputNode1),
      new OutputStream(outputNode2),
    ]);

    const result = convertFromDag(dag);
    expect(result).toMatchSnapshot();
  });

  it('should handle audio and video streams correctly', () => {
    // Create a DAG with both audio and video streams
    const inputNode = new InputNode('input.mp4');
    const videoFilter = new FilterNode(
      'reverse',
      [new VideoStream(inputNode)],
      [new StreamType('video')],
      [new StreamType('video')]
    );
    const audioFilter = new FilterNode(
      'areverse',
      [new AudioStream(inputNode)],
      [new StreamType('audio')],
      [new StreamType('audio')]
    );
    const outputNode = new OutputNode('output.mp4', [
      new VideoStream(videoFilter),
      new AudioStream(audioFilter),
    ]);
    const dag = new GlobalNode([new OutputStream(outputNode)]);

    const result = convertFromDag(dag);
    expect(result).toMatchSnapshot();
  });

  it('should preserve filter parameters', () => {
    // Create a DAG with a filter that has parameters
    const inputNode = new InputNode('input.mp4');
    const filterNode = new FilterNode(
      'scale',
      [new VideoStream(inputNode)],
      [new StreamType('video')],
      [new StreamType('video')],
      { width: 1920, height: 1080 }
    );
    const outputNode = new OutputNode('output.mp4', [new VideoStream(filterNode)]);
    const dag = new GlobalNode([new OutputStream(outputNode)]);

    const result = convertFromDag(dag);
    expect(result).toMatchSnapshot();
  });
}); 