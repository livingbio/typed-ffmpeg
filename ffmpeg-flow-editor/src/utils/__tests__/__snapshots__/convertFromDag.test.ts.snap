// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertFromDag > should convert a DAG with filter nodes to React Flow nodes and edges 1`] = `
{
  "edges": [
    {
      "data": {
        "type": "video",
      },
      "id": "edge-output-0-output-0-output-0-input-0",
      "source": "output-0",
      "sourceHandle": "output-0",
      "style": {
        "stroke": "#f44336",
      },
      "target": "output-0",
      "targetHandle": "input-0",
      "type": "smoothstep",
    },
  ],
  "nodes": [
    {
      "data": {
        "filename": "output.mp4",
        "filterString": "[outv]",
        "filterType": "output",
        "handles": {
          "inputs": [
            {
              "id": "input-0",
              "type": "av",
            },
          ],
          "outputs": [],
        },
        "label": "Output",
        "parameters": {},
      },
      "id": "output-0",
      "position": {
        "x": 800,
        "y": 100,
      },
      "type": "output",
    },
  ],
}
`;

exports[`convertFromDag > should convert a simple input-output DAG to React Flow nodes and edges 1`] = `
{
  "edges": [
    {
      "data": {
        "type": "video",
      },
      "id": "edge-output-0-output-0-output-0-input-0",
      "source": "output-0",
      "sourceHandle": "output-0",
      "style": {
        "stroke": "#f44336",
      },
      "target": "output-0",
      "targetHandle": "input-0",
      "type": "smoothstep",
    },
  ],
  "nodes": [
    {
      "data": {
        "filename": "output.mp4",
        "filterString": "[outv]",
        "filterType": "output",
        "handles": {
          "inputs": [
            {
              "id": "input-0",
              "type": "av",
            },
          ],
          "outputs": [],
        },
        "label": "Output",
        "parameters": {},
      },
      "id": "output-0",
      "position": {
        "x": 800,
        "y": 100,
      },
      "type": "output",
    },
  ],
}
`;

exports[`convertFromDag > should handle audio and video streams correctly 1`] = `
{
  "edges": [
    {
      "data": {
        "type": "video",
      },
      "id": "edge-output-0-output-0-output-0-input-0",
      "source": "output-0",
      "sourceHandle": "output-0",
      "style": {
        "stroke": "#f44336",
      },
      "target": "output-0",
      "targetHandle": "input-0",
      "type": "smoothstep",
    },
    {
      "data": {
        "type": "audio",
      },
      "id": "edge-output-0-output-0-output-0-input-1",
      "source": "output-0",
      "sourceHandle": "output-0",
      "style": {
        "stroke": "#2196f3",
      },
      "target": "output-0",
      "targetHandle": "input-1",
      "type": "smoothstep",
    },
  ],
  "nodes": [
    {
      "data": {
        "filename": "output.mp4",
        "filterString": "[outv]",
        "filterType": "output",
        "handles": {
          "inputs": [
            {
              "id": "input-0",
              "type": "av",
            },
            {
              "id": "input-1",
              "type": "av",
            },
          ],
          "outputs": [],
        },
        "label": "Output",
        "parameters": {},
      },
      "id": "output-0",
      "position": {
        "x": 800,
        "y": 100,
      },
      "type": "output",
    },
  ],
}
`;

exports[`convertFromDag > should handle multiple output nodes 1`] = `
{
  "edges": [
    {
      "data": {
        "type": "video",
      },
      "id": "edge-output-0-output-0-output-0-input-0",
      "source": "output-0",
      "sourceHandle": "output-0",
      "style": {
        "stroke": "#f44336",
      },
      "target": "output-0",
      "targetHandle": "input-0",
      "type": "smoothstep",
    },
    {
      "data": {
        "type": "video",
      },
      "id": "edge-output-0-output-1-output-0-input-0",
      "source": "output-0",
      "sourceHandle": "output-1",
      "style": {
        "stroke": "#f44336",
      },
      "target": "output-0",
      "targetHandle": "input-0",
      "type": "smoothstep",
    },
  ],
  "nodes": [
    {
      "data": {
        "filename": "output1.mp4",
        "filterString": "[outv]",
        "filterType": "output",
        "handles": {
          "inputs": [
            {
              "id": "input-0",
              "type": "av",
            },
          ],
          "outputs": [],
        },
        "label": "Output",
        "parameters": {},
      },
      "id": "output-0",
      "position": {
        "x": 800,
        "y": 100,
      },
      "type": "output",
    },
  ],
}
`;

exports[`convertFromDag > should preserve filter parameters 1`] = `
{
  "edges": [
    {
      "data": {
        "type": "video",
      },
      "id": "edge-output-0-output-0-output-0-input-0",
      "source": "output-0",
      "sourceHandle": "output-0",
      "style": {
        "stroke": "#f44336",
      },
      "target": "output-0",
      "targetHandle": "input-0",
      "type": "smoothstep",
    },
  ],
  "nodes": [
    {
      "data": {
        "filename": "output.mp4",
        "filterString": "[outv]",
        "filterType": "output",
        "handles": {
          "inputs": [
            {
              "id": "input-0",
              "type": "av",
            },
          ],
          "outputs": [],
        },
        "label": "Output",
        "parameters": {},
      },
      "id": "output-0",
      "position": {
        "x": 800,
        "y": 100,
      },
      "type": "output",
    },
  ],
}
`;
